### ISSUE TITLE
Strange heap-buffer-overflow when use afl-cc + ASan + UBSan

### ISSUE THREAD

**Issue by nataraj-hates-MS-for-stealing-github** [AFLplusplus/AFLplusplus#630]
Hi!

I met an unexpected behavior, while using afl-cc + ASan + UBSan. It is reproducible only in this combination, and disappears when you change something (change afl-cc to clang-11 or remove one of the sanitizers). I guess this behavior might be pointing to some bug.

# How to reproduce
My afl-cc is built with clang-11 that came from Debian clang-11 source package.

1. Fetch postgres code
```
git clone git://git.postgresql.org/git/postgresql.git -b REL_11_9
```

2. Apply this patch [ubsan_empty_memcpy_fix_11.diff.gz](https://github.com/AFLplusplus/AFLplusplus/files/5684563/ubsan_empty_memcpy_fix_11.diff.gz) that makes UBSan happy in several non-critical cases.

3. Configure source to use afl-cc and UBSan and ASan with ```-fsanitize=address -fsanitize=undefined -fsanitize-undefined-trap-on-
error``` options.

```
CC=afl-cc CXX=afl-cc  CFLAGS="-fsanitize=address -fsanitize=undefined -fsanitize-undefined-trap-on-
error" LDFLAGS="-fsanitize=address  -fsanitize=undefined  -fsanitize-undefined-trap-on-
error" ./configure --prefix=/tmp/pg 
```
4. Build and install
```
make
make install
```
5. Run initdb script
```
ASAN_OPTIONS=detect_leaks=0 /tmp/pg/bin/initdb -D /tmp/pg-data
```

6. Get error
```
=================================================================
==10743==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7fffe85d38d8 at pc 0x7fffe872823d bp 0x7fffffffc0d0 sp 0x7fffffffc0c8
READ of size 1 at 0x7fffe85d38d8 thread T0
    #0 0x7fffe872823c in shift_jis_20042euc_jis_2004 /tmp/eee/postgresql/src/backend/utils/mb/conversion_procs/euc2004_sjis2004/euc2004_sjis2004.c:332:5
    #1 0x7fffe872823c in shift_jis_2004_to_euc_jis_2004 /tmp/eee/postgresql/src/backend/utils/mb/conversion_procs/euc2004_sjis2004/euc2004_sjis2004.c:59:2
    #2 0x37cfec2 in OidFunctionCall5Coll /tmp/eee/postgresql/src/backend/utils/fmgr/fmgr.c:1533:11
    #3 0x12bdd2f in CreateConversionCommand /tmp/eee/postgresql/src/backend/commands/conversioncmds.c:101:2
    #4 0x2ad7c33 in ProcessUtilitySlow /tmp/eee/postgresql/src/backend/tcop/utility.c:1537:15
    #5 0x2acddd4 in standard_ProcessUtility /tmp/eee/postgresql/src/backend/tcop/utility.c
    #6 0x2acaa19 in ProcessUtility /tmp/eee/postgresql/src/backend/tcop/utility.c:360:3
    #7 0x2ac7391 in PortalRunUtility /tmp/eee/postgresql/src/backend/tcop/pquery.c:1178:2
    #8 0x2ac11ad in PortalRunMulti /tmp/eee/postgresql/src/backend/tcop/pquery.c
    #9 0x2abbd04 in PortalRun /tmp/eee/postgresql/src/backend/tcop/pquery.c:799:5
    #10 0x2aabba2 in exec_simple_query /tmp/eee/postgresql/src/backend/tcop/postgres.c:1145:10
    #11 0x2a95885 in PostgresMain /tmp/eee/postgresql/src/backend/tcop/postgres.c
    #12 0x1cd6d60 in main /tmp/eee/postgresql/src/backend/main/main.c:224:3
    #13 0x7ffff7c5709a in __libc_start_main /build/glibc-vjB4T1/glibc-2.28/csu/../csu/libc-start.c:308:16
    #14 0x4a2b69 in _start (/tmp/pg/bin/postgres+0x4a2b69)

0x7fffe85d38d8 is located 0 bytes to the right of 319704-byte region [0x7fffe8585800,0x7fffe85d38d8)
allocated by thread T0 here:
    #0 0x51cbdd in malloc (/tmp/pg/bin/postgres+0x51cbdd)
    #1 0x3b311a8 in __afl_map_shm llvm_mode/instrumentation/afl-compiler-rt.o.c:375:24
    #2 0x3b311a8 in __afl_auto_early llvm_mode/instrumentation/afl-compiler-rt.o.c:940:3

SUMMARY: AddressSanitizer: heap-buffer-overflow /tmp/eee/postgresql/src/backend/utils/mb/conversion_procs/euc2004_sjis2004/euc2004_sjis2004.c:332:5 in shift_jis_20042euc_jis_2004
Shadow bytes around the buggy address:
  0x10007d0b26c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007d0b26d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007d0b26e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007d0b26f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007d0b2700: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10007d0b2710: 00 00 00 00 00 00 00 00 00 00 00[fa]fa fa fa fa
  0x10007d0b2720: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x10007d0b2730: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x10007d0b2740: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x10007d0b2750: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x10007d0b2760: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==10743==ABORTING
```
If you change ```afl-cc``` to ```clang-11```, remove any of sanitzers, or remove ```-fsanitize-undefined-trap-on-error``` option, then this error message would disappear.

I asked my colleges to check that place in the postgres code, if there might be some dubious problem. They claimed that there everything is ok. So I guess it is some afl-cc issue.

May be there is some bug that can be tracked down using this case. So I report it.

**Comment by vanhauser-thc** [comment]
can you please generate a core, and use gdb to do a backtrace and disassemble from 5 instructions before the crash address up to 5 instructions after the crash address - and paste that here?

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
```shift_jis_20042euc_jis_2004``` begins with a loop
```
B+ │225             while (len > 0)
   │226             {
```
that is skipped (as len ==0) and this loop shown as asm code
```
B+ │0x7fffe8726e51 <shift_jis_2004_to_euc_jis_2004+337>     test   %ebx,%ebx
   │0x7fffe8726e53 <shift_jis_2004_to_euc_jis_2004+339>     jle    0x7fffe8728126 <shift_jis_2004_to_euc_jis_2004+5158>  
```
then it jumps to 
```
  >│332             *p = '\0';   
```
That is compiled into these
```
  >│0x7fffe872817c <shift_jis_2004_to_euc_jis_2004+5244>    test   %r13,%r13                                                                                                                   │
   │0x7fffe872817f <shift_jis_2004_to_euc_jis_2004+5247>    je     0x7fffe8728e2a <shift_jis_2004_to_euc_jis_2004+8490>                                                                        │
   │0x7fffe8728185 <shift_jis_2004_to_euc_jis_2004+5253>    lea    0x5298(%rip),%rax        # 0x7fffe872d424                                                                                   │
   │0x7fffe872818c <shift_jis_2004_to_euc_jis_2004+5260>    shr    $0x3,%rax                                                                                                                   │
   │0x7fffe8728190 <shift_jis_2004_to_euc_jis_2004+5264>    mov    0x7fff8000(%rax),%al                                                                                                        │
   │0x7fffe8728196 <shift_jis_2004_to_euc_jis_2004+5270>    test   %al,%al                                                                                                                     │
   │0x7fffe8728198 <shift_jis_2004_to_euc_jis_2004+5272>    jne    0x7fffe872820b <shift_jis_2004_to_euc_jis_2004+5387>                                                                        │
   │0x7fffe872819a <shift_jis_2004_to_euc_jis_2004+5274>    cmpb   $0x0,0x7fff8000(%r12)                                                                                                       │
   │0x7fffe87281a3 <shift_jis_2004_to_euc_jis_2004+5283>    jne    0x7fffe8728e67 <shift_jis_2004_to_euc_jis_2004+8551>                                                                        │
   │0x7fffe87281a9 <shift_jis_2004_to_euc_jis_2004+5289>    movslq 0x5274(%rip),%rdi        # 0x7fffe872d424                                                                                   │
   │0x7fffe87281b0 <shift_jis_2004_to_euc_jis_2004+5296>    add    (%r9),%rdi                                                                                                                  │
   │0x7fffe87281b3 <shift_jis_2004_to_euc_jis_2004+5299>    mov    %rdi,%rax                                                                                                                   │
   │0x7fffe87281b6 <shift_jis_2004_to_euc_jis_2004+5302>    shr    $0x3,%rax                                                                                                                   │
   │0x7fffe87281ba <shift_jis_2004_to_euc_jis_2004+5306>    mov    0x7fff8000(%rax),%al                                                                                                        │
  >│0x7fffe87281c0 <shift_jis_2004_to_euc_jis_2004+5312>    test   %al,%al                                                                                                                     │
   │0x7fffe87281c2 <shift_jis_2004_to_euc_jis_2004+5314>    jne    0x7fffe872822f <shift_jis_2004_to_euc_jis_2004+5423>                                                                        │
^^^^^^^ Here it jumps

   │0x7fffe87281c4 <shift_jis_2004_to_euc_jis_2004+5316>    mov    (%rdi),%al   

   │0x7fffe87281c6 <shift_jis_2004_to_euc_jis_2004+5318>    add    $0x1,%al                                                                                                                    │
   │0x7fffe87281c8 <shift_jis_2004_to_euc_jis_2004+5320>    adc    $0x0,%al                                                                                                                    │
   │0x7fffe87281ca <shift_jis_2004_to_euc_jis_2004+5322>    mov    %al,(%rdi)                                                                                                                  │
   │0x7fffe87281cc <shift_jis_2004_to_euc_jis_2004+5324>    mov    %r13,%rax                                                                                                                   │
   │0x7fffe87281cf <shift_jis_2004_to_euc_jis_2004+5327>    shr    $0x3,%rax               

..................
vvvvvvvvvvvvvvvvv jumps here 
 >│0x7fffe872822f <shift_jis_2004_to_euc_jis_2004+5423>    mov    %edi,%ecx                                                                                                                   │
   │0x7fffe8728231 <shift_jis_2004_to_euc_jis_2004+5425>    and    $0x7,%cl                                                                                                                    │
   │0x7fffe8728234 <shift_jis_2004_to_euc_jis_2004+5428>    cmp    %al,%cl                                                                                                                     │
   │0x7fffe8728236 <shift_jis_2004_to_euc_jis_2004+5430>    jl     0x7fffe87281c4 <shift_jis_2004_to_euc_jis_2004+5316>                                                                        │
   │0x7fffe8728238 <shift_jis_2004_to_euc_jis_2004+5432>    callq  0x7fffe87241c0 <__asan_report_load1@plt>                                                                                    │
^^^^^^^^^^^^^^^^^^^^ This prints error report.
   │0x7fffe872823d <shift_jis_2004_to_euc_jis_2004+5437>    mov    %r13d,%ecx                                                                                                                  │
   │0x7fffe8728240 <shift_jis_2004_to_euc_jis_2004+5440>    and    $0x7,%cl                                                                                                                    │
```

I've stepped with ```ni``` through it. Marking jumps with ^^^^^^^^^ and vvvvvvvvvvv. Hope it will help

**Comment by vanhauser-thc** [comment]
hhmmm this disassembly does not seem like the target was instrumented with afl-clang-fast (or any variant)

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
That's what I did 
```
CC=afl-cc CXX=afl-cc  CFLAGS="-fsanitize=address -fsanitize=undefined -fsanitize-undefined-trap-on-error" LDFLAGS="-fsanitize=address  -fsanitize=undefined  -fsanitize-undefined-trap-on-error" ./configure --prefix=/tmp/pg 
```

and the binary carps
```
Warning: AFL++ tools will need to set AFL_MAP_SIZE to 319184 to be able to run this instrumented program!
```
all around.
So I guess it is built with afl-cc O_o

**Comment by vanhauser-thc** [comment]
have you tried to fuzz it with afl-fuzz and AFL_MAP_SIZE=318184? are there crashes reported too?
and what is te output if you just execute `AFL_DEBUG=1 ASAN_OPTIONS=detect_leaks=0 /tmp/pg/bin/initdb -D /tmp/pg-data` ?

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
> have you tried to fuzz it with afl-fuzz 

No, I have not. initdb should be done before fuzzing. we actually do not fuzz it. 

But setting AFL_MAP_SIZE to some reasonable value does not help too. I've checked.

> and what is te output if you just execute AFL_DEBUG=1 ASAN_OPTIONS=detect_leaks=0 /tmp/pg/bin/initdb -D /tmp/pg-data ?

I've executed as you asked (not initdb, but initdb executes postres binary, and I've executed it as initdb do)

Several debug lines before crash are:

```
Running __sanitizer_cov_trace_pc_guard_init: 0x7fffe8b5d3d4-0x7fffe8b5d420
DEBUG: id_str <null>, __afl_area_ptr 0x7fffe89ca800, __afl_area_initial 0x4bcad20, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 319568, max_size_forkserver 8388608/0x800000
DEBUG: cmplog id_str <null>
Running __sanitizer_cov_trace_pc_guard_init: 0x7fffe89393d4-0x7fffe8939420
DEBUG: id_str <null>, __afl_area_ptr 0x7fffe87be800, __afl_area_initial 0x4bcad20, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 319592, max_size_forkserver 8388608/0x800000
DEBUG: cmplog id_str <null>
Running __sanitizer_cov_trace_pc_guard_init: 0x7fffe872d264-0x7fffe872d428
DEBUG: id_str <null>, __afl_area_ptr 0x7fffe85d6800, __afl_area_initial 0x4bcad20, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 319704, max_size_forkserver 8388608/0x800000
DEBUG: cmplog id_str <null>
```
full log can be found here
[fulllog.txt.gz](https://github.com/AFLplusplus/AFLplusplus/files/5708841/fulllog.txt.gz)

**Comment by vanhauser-thc** [comment]
OK, there something goes very wrong. it looks like afl-compiler-rt.o got multiple times linked, because that "DEBUG: id_str ..." should only be printed once, but in your output this happens mutliple times. or does that program execute other instrumented programs?

also try adding "AFL_DISABLE_LLVM_INSTRUMENTATION=1" when executing the program

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
> or does that program execute other instrumented programs?
No, it does not run anything.

> also try adding "AFL_DISABLE_LLVM_INSTRUMENTATION=1" when executing the program
```
AFL_DISABLE_LLVM_INSTRUMENTATION=1 AFL_MAP_SIZE=419704 AFL_DEBUG=1 ASAN_OPTIONS=detect_leaks=0 /tmp/pg/bin/postgres --single -F -O -j -c search_path=pg_catalog -c exit_on_error=true template1 >/dev/null -D /tmp/pg-data < /tmp/pg/post-bootstrap-init.bki
Running __sanitizer_cov_trace_pc_guard_init: 0x41515d4-0x42880d0
AddressSanitizer:DEADLYSIGNAL
=================================================================
==22039==ERROR: AddressSanitizer: SEGV on unknown address 0x000004c180f7 (pc 0x000003adbe0a bp 0x00000000000c sp 0x7fffffffe330 T0)
==22039==The signal is caused by a READ memory access.
    #0 0x3adbe0a in get_progname /tmp/ddd/postgresql/src/port/path.c:454
    #1 0x1cd519f in main /tmp/ddd/postgresql/src/backend/main/main.c:72:13
    #2 0x7ffff7c5709a in __libc_start_main /build/glibc-vjB4T1/glibc-2.28/csu/../csu/libc-start.c:308:16
    #3 0x4a2b69 in _start (/tmp/pg/bin/postgres+0x4a2b69)


AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /tmp/ddd/postgresql/src/port/path.c:454 in get_progname
==22039==ABORTING
```

**Comment by vanhauser-thc** [comment]
can you please retry with the current dev state? I think/hope it is fixed

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
Yes I guess it is fixed.

I tried 39a4fac941177387578ec856aacea2187588fc13 (Wed Dec 9 11:07:14 2020 +0100) bug is there,
ce9e127529f3ef43d826d3ec17a5b68b99f205da (Sat Dec 12 22:56:09 2020 +0100) bug is not there any more
e98cd008222aa3bfea9b696ad756163302437eb3 (Fri Mar 19 19:04:59 2021) -- current stable, bug is not there too...

Actually I do not know what was wrong, but was reproducible, and is gone now.
If it is important I can find a commit when it was fixed. (there are not many of them between 9 and 12 dec) if not, let's just close a bug.
